### ISSUE TITLE
Feature Request: Sorting in Add Weapons/Specials Popups on the Ship Design Screen

### ISSUE THREAD

**Issue by sytnik-aa** [1oom-fork/1oom#44]
I think it would be very useful if I could sort items in the add weapons and the add special devices popup windows by clicking on the table header (as I can do the same on the planets interface). Thanks!

**Comment by 1oom-fork** [comment]
This is a very important issue. I've been toying with the idea of ​​adding weapon categories such as missiles, bombs, energy or kinetic weapons. It will be necessary to make major changes to the interface and I’m not yet sure that my idea will be implemented beautifully, so I can try to do this only after I deal with the problems in the kernel, or if there are no more important problems that I want to deal with in the meantime moment.

**Comment by sytnik-aa** [comment]
Just wanted to add some screens there. 

In the planets interface there are no addional interface helpers for the sorting, you just click on the table's header and it works: 

<img width="703" alt="image" src="https://github.com/1oom-fork/1oom/assets/40230135/9d40db5f-2bb9-4b3e-bc37-777ab3b2658b">

I think that's OK.

So for the add weapons (special, computer, shield, etc.) popup it could work similar -- just click on the header, no visual indication, no addional interface changes.

<img width="716" alt="image" src="https://github.com/1oom-fork/1oom/assets/40230135/c976103e-4f94-40a6-a895-93df11678527">

The idea of groups is really nice (there are, actuallty three weapon groups mentione in the game's manual: Beams, Bombs & Biological and Missiles & Torpedos) and it would be cool to have them too, but you have to add something to the interface which is a big deal. 

For example "Weapons" header could renamed to "All Weapons" and converted into a carousel that allows you to select a weapon type you prefer using arrows like these: 

<img width="29" alt="image" src="https://github.com/1oom-fork/1oom/assets/40230135/0316d44a-e61c-4085-aafb-fe9803785778">

![image](https://github.com/1oom-fork/1oom/assets/40230135/ab370cb5-e346-4f8a-ac4c-4115f05c92f1)

Just an idea.

**Comment by 1oom-fork** [comment]
I like it. In any case, I will consider my options, and changes can always be reversed. I like the illustration, it's quite doable. Thanks for your help with the design. This certainly simplifies the task.

**Comment by 1oom-fork** [comment]
Replacing the caption is trivial since the caption uses the game font. Rotating the arrows is either already implemented, or can be implemented similar to already implemented transformations. All that remains is to write the filter. I'll try to do it in the next few days

**Comment by 1oom-fork** [comment]
game_design_build_tbl_fit_weapon is responsible for filtering. Now the task is almost trivial

**Comment by 1oom-fork** [comment]
The central problem is that in the tables there is no such thing as beam or missile. Instead, some similarities that have already been clarified are used to determine the type. It follows that changing the characteristics of weapons using PBX can potentially break the behavior

**Comment by 1oom-fork** [comment]
To eliminate the problem mentioned above, it is enough to accept as a rule that Beam Weapons can fire an unlimited number of times. Is this specified somewhere in the OSG?

**Comment by sytnik-aa** [comment]
What pieces of information, other than we see on the screen, are available to you in code? Could dump it, please I'll analyze it.

**Comment by 1oom-fork** [comment]
We have explicit flags is_bomb, is_bio, halveshield (kinetic beams), as well as the torpedo's appearance number. In addition, there is a variable with the number of volleys (for example, a Gatling laser), multiple damage (Hellfire Torpedo, Stellar Converter) and the number of substances (Scatter Pack, Graviton Beam (= 1)), but the most ambiguous is the number of volleys (always -1 for beams)

**Comment by 1oom-fork** [comment]
I still managed to find the correct way to identify a beam weapon. is_missile = ((w->damagemax == w->damagemin) && (!w->is_bomb) && (w->misstype == 0)); This literally means the following: A missile is not a bomb, not a torpedo and a weapon with constant damage. The maximum number of shots is not used, although it looks like the simplest sign for beams

**Comment by 1oom-fork** [comment]
I'm waiting for your opinion. The plan is ready, all that remains is to complicate the behavior of the lists a little, this is already a routine.

**Comment by 1oom-fork** [comment]
When the complex behavior of weapon lists is implemented, it will be possible to use this mechanism to also separate, for example, colony modules from combat modules

**Comment by sytnik-aa** [comment]
I'm sorry, I can't review the logic right now, but I have a question. In order to properly test the feature, it is good to have access to all weapons and ship specials (in other words, have all technologies researched or traded/stolen). But how to achieve that? Psilon has access to more technology than other races (but I don't know what that means in numbers), but it's not enough. Is there a way to make all technologies researchable?

**Comment by 1oom-fork** [comment]
You can create a game that has all the technology ready. If you know how to hack using saveconv

**Comment by 1oom-fork** [comment]
I'll clean it up a little later and write it more clearly.

**Comment by 1oom-fork** [comment]
I cleaned up a bit of the mess I made here. I'm currently working on a critical issue in the core of the game. The importance of correcting this error is invaluable, but it requires significant effort. I will be forced to postpone the implementation of minor proposals, but this does not mean that I refuse them. Just don't expect it in the next week or two.

**Comment by 1oom-fork** [comment]
I found a few hours to implement a test version. You need to enable ui-extra and click on Weapons to change the filter. I'll post the binaries later if necessary.
fb51d4b9ff5904a8026558f5ef8494298125bbff
7adf5054f91f8443333b001d98623331f843516d

**Comment by sytnik-aa** [comment]
Yes, please, post the binaries if you have time to build them. Thanks!

**Comment by 1oom-fork** [comment]
There is time, but I would be very happy if the people who are testing the project knew how to assemble it themselves.
https://github.com/1oom-fork/1oom/releases/download/f1.8.1/1oom-fork-v1.8.1-x64-2.zip

**Comment by 1oom-fork** [comment]
In fact, the groups that I wrote there can be changed as you please. There is no problem separating fast and slow missiles, light and heavy lasers. Maybe I'll add this later.

**Comment by sytnik-aa** [comment]
Thank you! Much appreciated! Next time I'll try to compile it myself! 

I've tested the build! It's awesome! I did a lot of clicking in various combinations and I didn't encounter any problems. 

The only flaw I see is that the popup can have a different height depending on how many elements of a specific group you have, so it keeps changing its position, which is not great. Maybe the popup should have its position fixed.

**Comment by 1oom-fork** [comment]
Yes, I didn't like it either.

**Comment by 1oom-fork** [comment]
I was thinking, you can, of course, try to limit the minimum height, but you won’t be able to limit the maximum

**Comment by 1oom-fork** [comment]
There are too many options on how to fix this. I'm closing this question for now
